bofire/strategies/strategy.py

```diff
@@ -0,0 +1,376 @@
+        """Validator to ensure that all constraints defined in the domain are valid for the chosen strategy
+
+        Args:
+            domain (Domain): The domain to be used in the strategy
+
+        Raises:
+            ValueError: if a constraint is defined in the domain but is invalid for the strategy chosen
+
+        Returns:
+            Domain: the domain
+        """
+        for constraint in domain.constraints:
+            if not cls.is_constraint_implemented(type(constraint)):
+                raise ValueError(
+                    f"constraint `{type(constraint)}` is not implemented for strategy `{cls.__name__}`"
+                )
+        return domain
+
+    @validator("domain")
+    def validate_features(cls, domain: Domain):
+        """Validator to ensure that all features defined in the domain are valid for the chosen strategy
+
+        Args:
+            domain (Domain): The domain to be used in the strategy
+
+        Raises:
+            ValueError: if a feature type is defined in the domain but is invalid for the strategy chosen
+
+        Returns:
+            Domain: the domain
+        """
+        for feature in domain.input_features + domain.output_features:
+            if not cls.is_feature_implemented(type(feature)):
+                raise ValueError(
+                    f"feature `{type(feature)}` is not implemented for strategy `{cls.__name__}`"
+                )
+        return domain
+
+    @validator("domain")
+    def validate_objectives(cls, domain: Domain):
+        """Validator to ensure that all objectives defined in the domain are valid for the chosen strategy
+
+        Args:
+            domain (Domain): The domain to be used in the strategy
+
+        Raises:
+            ValueError: if a objective type is defined in the domain but is invalid for the strategy chosen
+
+        Returns:
+            Domain: the domain
+        """
+        for feature in domain.get_outputs_by_objective(Objective):
+            if not cls.is_objective_implemented(type(feature.objective)):
+                raise ValueError(
+                    f"Objective `{type(feature)}` is not implemented for strategy `{cls.__name__}`"
+                )
+        return domain
+
+    @abstractmethod
+    def _init_domain(
+        self,
+    ) -> None:
+        """Abstract method to allow for customized functions in the constructor of Strategy.
+
+        Called at the end of `__init__`.
+        """
+        pass
+
+    @property
+    def experiments(self) -> pd.DataFrame:
+        """Property returning the experiments associated with the current strategy.
+
+        Returns:
+            pd.DataFrame: Experiments.
+        """
+        return self.domain.experiments
+
+    @property
+    def pending_candidates(self) -> pd.DataFrame:
+        """Candidates considered as pending.
+
+        Returns:
+            pd.DataFrame: pending candidates.
+        """
+        return self.domain.candidates
+
+    def tell(
```

Comment:
Why do algorithms without model such as random sampling or doe need a tell function?

Reply:
Even without model DoE-based methods depend on the location of points in input space.

Reply:
I agree, we will also need it for closed loop experimentation for benchmark problems in python there the order is always:
- tell(experiments)
- ask() --> candidates
- f(candidates) --> experiments
- 
But I agree also with Behrang, that there are strategies where it is not needed, for this purpose I will create a further subclass in which I overwrite _tell() with pass so that this does not has to be done in every strategy again.